### PR TITLE
[TSI] Add a sample for Time Series Insights hierarchy

### PR DESCRIPTION
**Pull request by bikamani**
Learn more about hierarchy here: https://docs.microsoft.com/en-us/azure/time-series-insights/concepts-model-overview#time-series-model-hierarchies

Changes to be merged in to remote branch for all sample changes.

PR had some merge conflicts and is resolved (even though on UI it shows differently) - only 2 files are modified/added.

### CODE REVIEW

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/Program.cs [hunk 1]
```diff
@@ -44,6 +44,9 @@ public static async Task Main(string[] args)
 
             var tsiTypesSamples = new TypesSamples();
             await tsiTypesSamples.RunSamplesAsync(tsiClient);
+
+            var tsiHiearchiesSamples = new HierarchiesSamples();
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/Program.cs:48]: misspelled hierarchies

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        public async Task RunSamplesAsync(TimeSeriesInsightsClient client)
+        {
+            // For the purpose of keeping code snippets readable to the user, hardcoded string literals are used in place of assigned variables, eg Ids.
+            // Despite not being a good code practice, this prevents code snippets from being out of context for the user when making API calls that accept Ids as parameters.
+
+            PrintHeader("TIME SERIES INSIGHTS HIERARCHIES SAMPLE");
+
+            #region Snippet:TimeSeriesInsightsSampleCreateHierarchies
+            var tsiHierarchyName = "sampleHierarchy";
+            var tsiInstanceField1 = "hiearchyLevel1";
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:26]: I admit, hierarchy is a tricky word to spell :)

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            var hierarchySource = new TimeSeriesHierarchySource();
+            hierarchySource.InstanceFieldNames.Add(tsiInstanceField1);
+
+            var tsiHierarchy = new TimeSeriesHierarchy(tsiHierarchyName, hierarchySource);
+            tsiHierarchy.Id = "sampleHiearchyId";
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:31]: Hierarchy

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                }
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleCreateHierarchies
+
+            #region Snippet:TimeSeriesInsightsSampleGetAllHiearchies
+            // Get all Time Series hierarchies in the environment
+            AsyncPageable<TimeSeriesHierarchy> getAllHierarchies = client.Hierarchies.GetAsync();
+            await foreach (TimeSeriesHierarchy hierarchy in getAllHierarchies)
+            {
+                Console.WriteLine($"Retrieved Time Series Insights hierarchy with Id: '{hierarchy?.Id}' and Name: '{hierarchy?.Name}'");
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:64]: why would hierarchy be null? Is that a possibility?

Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:64]: dot at end of line

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleGetAllHiearchies
+
+            #region Snippet:TimeSeriesInsightsSampleReplaceHiearchies
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:68]: Hierarchies

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            #endregion Snippet:TimeSeriesInsightsSampleCreateHierarchies
+
+            #region Snippet:TimeSeriesInsightsSampleGetAllHiearchies
+            // Get all Time Series hierarchies in the environment
+            AsyncPageable<TimeSeriesHierarchy> getAllHierarchies = client.Hierarchies.GetAsync();
+            await foreach (TimeSeriesHierarchy hierarchy in getAllHierarchies)
+            {
+                Console.WriteLine($"Retrieved Time Series Insights hierarchy with Id: '{hierarchy?.Id}' and Name: '{hierarchy?.Name}'");
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleGetAllHiearchies
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:66]: Hierarchies

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                    Console.WriteLine($"Created Time Series hierarchy successfully.");
+                }
+                else
+                {
+                    Console.WriteLine($"Failed to create a Time Series hieararchy: {createHierarchiesResult.Value[i].Error.Message}.");
+                }
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleCreateHierarchies
+
+            #region Snippet:TimeSeriesInsightsSampleGetAllHiearchies
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:59]: Hierarchies

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            // Get all Time Series hierarchies in the environment
+            AsyncPageable<TimeSeriesHierarchy> getAllHierarchies = client.Hierarchies.GetAsync();
+            await foreach (TimeSeriesHierarchy hierarchy in getAllHierarchies)
+            {
+                Console.WriteLine($"Retrieved Time Series Insights hierarchy with Id: '{hierarchy?.Id}' and Name: '{hierarchy?.Name}'");
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleGetAllHiearchies
+
+            #region Snippet:TimeSeriesInsightsSampleReplaceHiearchies
+            // Update hieararchies with adding a new instance field
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:69]: hierarchies

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            var tsiInstanceField2 = "hiearchyLevel2";
+            foreach (TimeSeriesHierarchy hieararchy in timeSeriesHierarchies)
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:71]: hieararchy

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+                    .CreateOrReplaceAsync(timeSeriesHierarchies)
+                    .ConfigureAwait(false);
+
+            // The response of calling the API contains a list of error objects corresponding by position to the input parameter array in the request.
+            // If the error object is set to null, this means the operation was a success.
+            for (int i = 0; i < updateHierarchiesResult.Value.Length; i++)
+            {
+                if (updateHierarchiesResult.Value[i].Error == null)
+                {
+                    Console.WriteLine($"Updated Time Series hieararchy successfully.");
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:87]: hierarchy

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+                {
+                    Console.WriteLine($"Failed to update a Time Series Insights hierarchy due to: {updateHierarchiesResult.Value[i].Error.Message}.");
+                }
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleReplaceHiearchies
+
+            #region Snippet:TimeSeriesInsightsSampleGetHiearchiesById
+            var tsiHierarchyIds = new List<string>
+            {
+                "sampleHiearchyId"
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:99]: sampleHierarchyId

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+                {
+                    Console.WriteLine($"Failed to retrieve a Time Series hieararchy due to '{getHierarchiesByIdsResult.Value[i].Error.Message}'.");
+                }
+            }
+            #endregion Snippet:TimeSeriesInsightsSampleGetHiearchiesById
+
+            // Clean up
+            try
+            {
+                #region Snippet:TimeSeriesInsightsSampleDeleteHiearchiesById
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:125]: Hierarchies

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+                };
+
+                Response<TimeSeriesOperationError[]> deleteHierarchiesResponse = await client
+                        .Hierarchies
+                        .DeleteByIdAsync(tsiHierarchyIdsToDelete)
+                        .ConfigureAwait(false);
+
+                // The response of calling the API contains a list of error objects corresponding by position to the input parameter
+                // array in the request. If the error object is set to null, this means the operation was a success.
+                foreach (var result in deleteHierarchiesResponse.Value)
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:139]: change from var to explicit type

**Comment by barustum** [comment]
I am assuming sample snippets will be added in a different PR?

**Comment by bikamani** [comment]
> s will be added in a differ

yes

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs (hunk 4 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:64]: Good observation. Couldn't recall a reason why I added this check, will fix.

File: sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs (hunk 2 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.IoT.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/HierarchiesSamples.cs:26]: I am sure you had to review after you typed as well 😆

**Pull request merged**